### PR TITLE
[Cocoa] Initial support for HDR images

### PR DESCRIPTION
**Pull request by shallawa**
#### 8c53cb365a2d5aec99483ae9d855472c8a720941
<pre>
[Cocoa] Initial support for HDR images
<a href="https://bugs.webkit.org/show_bug.cgi?id=280583">https://bugs.webkit.org/show_bug.cgi?id=280583</a>
<a href="https://rdar.apple.com/136926899">rdar://136926899</a>

Reviewed by Mike Wyrzykowski.

A conditional compilation directive flag named `HAVE_HDR_SUPPORT` is added.

An internal feature flag named `HDRForImagesEnabled` is added.

A new option named Headroom is added to ImagePaintingOptions.

When rendering to a CABackingStore, the layer.contentsFormat needs to be set to
float 16 bit format.

When creating an IOSurface this IOSurface will be rendered via layer.contents,
the color space of this IOSurface has to be CGColorSpaceUsesExtendedRange(). The
pixel format of this IOSurface has to be float 16 bit format.

When displaying CGImage, its headroom has to be set to the context by calling
CGContextSetEDRTargetHeadroom().

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WTF/wtf/PlatformHave.h:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/css/query/MediaQueryFeatures.cpp:
(WebCore::MQ::Features::colorGamut):
* Source/WebCore/page/PageOverlayController.cpp:
(WebCore::PageOverlayController::updateSettingsForLayer):
* Source/WebCore/platform/PlatformScreen.h:
* Source/WebCore/platform/ScreenProperties.h:
* Source/WebCore/platform/graphics/BitmapImage.cpp:
(WebCore::BitmapImage::draw):
* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageSource.cpp:
(WebCore::BitmapImageSource::frameHeadroomAtIndex const):
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/ContentsFormat.cpp: Copied from Source/WebKit/WebProcess/GPU/graphics/RemoteNativeImageBackendProxy.h.
(WebCore::contentsFormatExtendedColorSpace):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/ContentsFormat.h: Copied from Source/WebCore/platform/graphics/ca/cocoa/WebTiledBackingLayer.h.
(WebCore::contentsFormatBytesPerPixel):
* Source/WebCore/platform/graphics/GraphicsLayer.h:
(WebCore::GraphicsLayer::setHDRForImagesEnabled):
(WebCore::GraphicsLayer::hdrForImagesEnabled const):
* Source/WebCore/platform/graphics/GraphicsLayerClient.h:
(WebCore::GraphicsLayerClient::hdrForImagesEnabled const):
* Source/WebCore/platform/graphics/ImageBufferPixelFormat.h:
(WebCore::convertToPixelFormat):
(WebCore::convertToIOSurfaceFormat):
* Source/WebCore/platform/graphics/ImageFrame.h:
(WebCore::ImageFrame::setHeadroom):
(WebCore::ImageFrame::headroom const):
* Source/WebCore/platform/graphics/ImagePaintingOptions.h:
(WebCore::ImagePaintingOptions::allowImageSubsampling const):
(WebCore::ImagePaintingOptions::showDebugBackground const):
(WebCore::ImagePaintingOptions::headroom const):
(WebCore::ImagePaintingOptions::setOption):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::frameHeadroomAtIndex const):
* Source/WebCore/platform/graphics/ImageTypes.h:
(WebCore::Headroom::Headroom):
(WebCore::Headroom::operator float const):
* Source/WebCore/platform/graphics/LayerTreeAsTextOptions.h:
* Source/WebCore/platform/graphics/NativeImage.cpp:
(WebCore::NativeImage::headroom const):
* Source/WebCore/platform/graphics/NativeImage.h:
* Source/WebCore/platform/graphics/PixelBuffer.cpp:
(WebCore::PixelBuffer::supportedPixelFormat):
* Source/WebCore/platform/graphics/PixelBufferConversion.cpp:
(WebCore::makeVImageCGImageFormat):
* Source/WebCore/platform/graphics/PixelFormat.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/PixelFormat.h:
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::createPlatformCALayer):
(WebCore::GraphicsLayerCA::dumpAdditionalProperties const):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.h:
* Source/WebCore/platform/graphics/ca/PlatformCALayer.mm:
(WebCore::PlatformCALayer::drawRepaintIndicator):
* Source/WebCore/platform/graphics/ca/PlatformCALayerClient.h:
(WebCore::PlatformCALayerClient::hdrForImagesEnabled const):
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::setContentsFormat):
(WebCore::TileController::createTileLayer):
(WebCore::TileController::setWantsDeepColorBackingStore): Deleted.
* Source/WebCore/platform/graphics/ca/TileController.h:
* Source/WebCore/platform/graphics/ca/TileGrid.cpp:
(WebCore::TileGrid::updateTileLayerProperties):
* Source/WebCore/platform/graphics/ca/cocoa/ContentsFormatCocoa.h: Copied from Source/WebCore/platform/graphics/ImageBufferPixelFormat.h.
(WebCore::convertToIOSurfaceFormat):
(WebCore::contentsFormatString):
(WebCore::contentsFormatWantsExtendedDynamicRangeContent):
(WebCore::contentsFormatWantsToneMapMode):
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:
* Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm:
(WebCore::PlatformCALayerCocoa::commonInit):
(WebCore::PlatformCALayerCocoa::contentsFormat const):
(WebCore::PlatformCALayerCocoa::setContentsFormat):
(WebCore::PlatformCALayerCocoa::updateContentsFormat):
(WebCore::PlatformCALayerCocoa::backingStoreBytesPerPixel const):
(WebCore::PlatformCALayerCocoa::wantsDeepColorBackingStore const): Deleted.
(WebCore::PlatformCALayerCocoa::setWantsDeepColorBackingStore): Deleted.
(WebCore::layerContentsFormat): Deleted.
* Source/WebCore/platform/graphics/ca/cocoa/WebTiledBackingLayer.h:
* Source/WebCore/platform/graphics/ca/cocoa/WebTiledBackingLayer.mm:
(-[WebTiledBackingLayer setContentsFormat:]):
(-[WebTiledBackingLayer contentsFormat]):
(-[WebTiledBackingLayer setWantsDeepColorBackingStore:]): Deleted.
(-[WebTiledBackingLayer wantsDeepColorBackingStore]): Deleted.
* Source/WebCore/platform/graphics/cairo/NativeImageCairo.cpp:
(WebCore::PlatformImageNativeImageBackend::headroom const):
* Source/WebCore/platform/graphics/cg/GraphicsContextCG.cpp:
(WebCore::GraphicsContextCG::drawNativeImageInternal):
* Source/WebCore/platform/graphics/cg/ImageDecoderCG.cpp:
(WebCore::createImageSourceOptions):
(WebCore::ImageDecoderCG::fetchFrameMetaDataAtIndex const):
* Source/WebCore/platform/graphics/cg/NativeImageCG.cpp:
(WebCore::PlatformImageNativeImageBackend::headroom const):
(WebCore::NativeImage::draw):
* Source/WebCore/platform/graphics/cocoa/IOSurface.h:
* Source/WebCore/platform/graphics/cocoa/IOSurface.mm:
(WebCore::optionsForSurface):
(WebCore::optionsFor32BitSurface):
(WebCore::optionsFor64BitSurface):
(WebCore::IOSurface::IOSurface):
(WebCore::formatFromSurface):
(WebCore::IOSurface::bitmapConfiguration const):
(WebCore::operator&lt;&lt;):
* Source/WebCore/platform/graphics/skia/NativeImageSkia.cpp:
(WebCore::PlatformImageNativeImageBackend::headroom const):
* Source/WebCore/platform/ios/LegacyTileGridTile.mm:
(WebCore::LegacyTileGridTile::LegacyTileGridTile):
* Source/WebCore/platform/ios/PlatformScreenIOS.mm:
(WebCore::screenContentsFormat):
(WebCore::screenColorSpace):
(WebCore::collectScreenProperties):
(WebCore::screenSupportsExtendedColor): Deleted.
* Source/WebCore/platform/mac/PlatformScreenMac.mm:
(WebCore::screenContentsFormat):
(WebCore::collectScreenProperties):
(WebCore::screenSupportsExtendedColor): Deleted.
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::hdrForImagesEnabled const):
* Source/WebCore/rendering/RenderLayerBacking.h:
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::hdrForImagesEnabled const):
* Source/WebCore/rendering/RenderLayerCompositor.h:
* Source/WebCore/testing/Internals.cpp:
(WebCore::toLayerTreeAsTextOptions):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:
* Source/WebKit/Shared/RemoteLayerTree/LayerProperties.h:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.h:
(WebKit::RemoteLayerBackingStore::contentsFormat const):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerBackingStore.mm:
(WebKit::RemoteLayerBackingStore::pixelFormat const):
(WebKit::RemoteLayerBackingStore::bytesPerPixel const):
(WebKit::RemoteLayerBackingStore::usesDeepColorBackingStore const): Deleted.
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTree.serialization.in:
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm:
(WebKit::RemoteLayerTreePropertyApplier::applyPropertiesToLayer):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreeTransaction.mm:
(WebKit::dumpChangedLayers):
* Source/WebKit/Shared/RemoteLayerTree/RemoteLayerWithRemoteRenderingBackingStore.mm:
(WebKit::RemoteLayerWithRemoteRenderingBackingStore::ensureBackingStore):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _takeViewSnapshot]):
* Source/WebKit/WebProcess/GPU/graphics/RemoteNativeImageBackendProxy.cpp:
(WebKit::RemoteNativeImageBackendProxy::headroom const):
* Source/WebKit/WebProcess/GPU/graphics/RemoteNativeImageBackendProxy.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/GraphicsLayerCARemote.mm:
(WebKit::GraphicsLayerCARemote::createPlatformCALayer):
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm:
(WebKit::PlatformCALayerRemote::displayColorSpace const):
(WebKit::PlatformCALayerRemote::updateBackingStore):
(WebKit::PlatformCALayerRemote::contentsFormat const):
(WebKit::PlatformCALayerRemote::setContentsFormat):
(WebKit::PlatformCALayerRemote::wantsDeepColorBackingStore const): Deleted.
(WebKit::PlatformCALayerRemote::setWantsDeepColorBackingStore): Deleted.
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteTiledBacking.cpp:
(WebKit::PlatformCALayerRemoteTiledBacking::contentsFormat const):
(WebKit::PlatformCALayerRemoteTiledBacking::setContentsFormat):
(WebKit::PlatformCALayerRemoteTiledBacking::wantsDeepColorBackingStore const): Deleted.
(WebKit::PlatformCALayerRemoteTiledBacking::setWantsDeepColorBackingStore): Deleted.
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemoteTiledBacking.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteLayerTreeDrawingArea.mm:
(WebKit::RemoteLayerTreeDrawingArea::hdrForImagesEnabled const):
* Tools/DumpRenderTree/ios/PixelDumpSupportIOS.mm:
(createBitmapContextFromWebView):

Canonical link: <a href="https://commits.webkit.org/284617@main">https://commits.webkit.org/284617@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/a54c7279a61f0b01d2fee39562b9af20571f1e94

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69852 "7 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49253 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22605 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73937 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21010 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71969 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57053 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20861 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/55464 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13936 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72918 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44885 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60252 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35945 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41550 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17686 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19387 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/62968 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63480 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18030 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75652 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/69098 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14077 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17267 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/63152 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14112 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60327 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63082 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15529 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11090 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4704 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90880 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45056 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/16144 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46130 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47401 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45871 "Built successfully") | | |

### CODE REVIEW

File: Source/WTF/wtf/PlatformHave.h [hunk 1]
```diff
@@ -971,6 +975,10 @@
 #endif
 #endif
 
+#if PLATFORM(IOS_FAMILY) && !PLATFORM(MACCATALYST) && !PLATFORM(IOS_FAMILY_SIMULATOR)
+#define HAVE_IOSURFACE_RGB10 1
```
Comment (mwyrzykowski) [inline Source/WTF/wtf/PlatformHave.h:979]: should we expose this for Apple Silicon macOS as well? RGB10 formats are supported there too

File: Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm [hunk 2]
```diff
@@ -1131,24 +1132,23 @@ - (void)setOwner:(WebCore::PlatformCALayer*)owner
...
 {
     if (m_layerType == PlatformCALayer::LayerType::LayerTypeWebLayer || m_layerType == PlatformCALayer::LayerType::LayerTypeTiledBackingTileLayer) {
         BEGIN_BLOCK_OBJC_EXCEPTIONS
-        if (NSString *formatString = layerContentsFormat(wantsDeepColorBackingStore()))
+        auto contentsFormat = this->contentsFormat();
+
+        if (NSString *formatString = contentsFormatToString(contentsFormat))
             [m_layer setContentsFormat:formatString];
+#if HAVE(HDR_SUPPORT)
+        if (contentsFormat == ContentsFormat::RGBA16F) {
```
Comment (mwyrzykowski) [inline Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm]: Should we also check for `ContentsFormat::RGBA10`? EDR is supported with that format too, right? The example from: https://developer.apple.com/documentation/metal/hdr_content/using_color_spaces_to_display_hdr_content?language=objc

uses `.wantsExtendedDynamicRangeContent = YES;` with `.pixelFormat = MTLPixelFormatBGR10A2Unorm;`

File: Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm [hunk 3]
```diff
@@ -325,6 +326,21 @@ static void updateCustomAppearance(CALayer *layer, GraphicsLayer::CustomAppearan
             [(WebAVPlayerLayer*)playerLayer setVideoGravity:convertMediaPlayerToAVLayerVideoGravity(properties.videoGravity)];
     }
 #endif
+
+    if (properties.changedProperties & LayerChange::ContentsFormatChanged) {
+        if (NSString *formatString = contentsFormatToString(properties.contentsFormat)) {
+            [layer setContentsFormat:formatString];
+#if HAVE(HDR_SUPPORT)
+            if (properties.contentsFormat == ContentsFormat::RGBA16F) {
```
Comment (mwyrzykowski) [inline Source/WebKit/Shared/RemoteLayerTree/RemoteLayerTreePropertyApplier.mm]: same comment as above regarding potentially considering `ContentsFormat::RGBA10`

File: Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm [hunk 4]
```diff
@@ -290,6 +290,35 @@
...
 
+DestinationColorSpace PlatformCALayerRemote::displayColorSpace() const
+{
+    switch (contentsFormat()) {
+    case ContentsFormat::RGBA8:
+        break;
+
+#if HAVE(IOSURFACE_RGB10)
+    case ContentsFormat::RGBA10:
+#if PLATFORM(IOS_FAMILY)
```
Comment (mwyrzykowski) [inline Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm]: should we use `HAVE(CORE_GRAPHICS_EXTENDED_SRGB_COLOR_SPACE)` instead or is there a reason it is iOS specific?

**Review by mwyrzykowski — APPROVED** [review]
Minor comments about RGB10 support on Apple Silicon Macs, but otherwise 👍

File: Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm (hunk 2 above)
Comment (squelart) [reply Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm]: And possibly relatedly: WebGPU allows the use of RGBA16F but without showing its extended dynamic range when displayed! https://www.w3.org/TR/webgpu/#gpucanvastonemappingmode
This tells me that maybe the contents format shouldn't directly force whether E/HDR should be shown? So we may actually need a separate bit for that.

File: Source/WebCore/platform/graphics/ca/PlatformCALayer.h [hunk 5]
```diff
@@ -210,8 +210,8 @@ class WEBCORE_EXPORT PlatformCALayer : public ThreadSafeRefCounted<PlatformCALay
     virtual bool acceleratesDrawing() const = 0;
     virtual void setAcceleratesDrawing(bool) = 0;
 
-    virtual bool wantsDeepColorBackingStore() const = 0;
-    virtual void setWantsDeepColorBackingStore(bool) = 0;
+    virtual ContentsFormat contentsFormat() const = 0;
+    virtual void setContentsFormat(ContentsFormat) = 0;
```
Comment (squelart) [inline Source/WebCore/platform/graphics/ca/PlatformCALayer.h:215]: You may need to forward-declare `ContentsFormat` near the top of this file (unless it's somehow already known through an include?)

File: Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h [hunk 6]
```diff
@@ -231,7 +231,7 @@ class PlatformCALayerCocoa final : public PlatformCALayer {
     ScrollingNodeID m_scrollingNodeID;
 #endif
     EventRegion m_eventRegion;
-    bool m_wantsDeepColorBackingStore { false };
+    ContentsFormat m_contentsFormat { ContentsFormat::RGBA8 };
```
Comment (squelart) [inline Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:235]: You may need to #include ContentsFormat.h.

Comment (shallawa) [reply Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.h:235]: The ContentsFormat.h was included.

File: Source/WebCore/platform/graphics/ca/PlatformCALayer.h (hunk 5 above)
Comment (shallawa) [reply Source/WebCore/platform/graphics/ca/PlatformCALayer.h:215]: The forward declaration is added.

File: Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm (hunk 4 above)
Comment (shallawa) [reply Source/WebKit/WebProcess/WebPage/RemoteLayerTree/PlatformCALayerRemote.mm]: We have this definition for `HAVE_IOSURFACE_RGB10`
```
#if PLATFORM(IOS_FAMILY) && !PLATFORM(MACCATALYST) && !PLATFORM(IOS_FAMILY_SIMULATOR)
#define HAVE_IOSURFACE_RGB10 1
#endif
```
And this definition for `HAVE_CORE_GRAPHICS_EXTENDED_SRGB_COLOR_SPACE`
```
#if PLATFORM(COCOA)
#define HAVE_CORE_GRAPHICS_EXTENDED_SRGB_COLOR_SPACE 1
#endif
```

So there is no need for either `PLATFORM(IOS_FAMILY)` or `HAVE(CORE_GRAPHICS_EXTENDED_SRGB_COLOR_SPACE)`

File: Source/WTF/wtf/PlatformHave.h (hunk 1 above)
Comment (shallawa) [reply Source/WTF/wtf/PlatformHave.h:979]: I just moved this definition from IOSurface.h to PlatformHave.h.

When changing this definition to be 
```
#if (PLATFORM(MAC) && CPU(ARM64)) || (PLATFORM(IOS_FAMILY) && !PLATFORM(MACCATALYST) && !PLATFORM(IOS_FAMILY_SIMULATOR))
#define HAVE_IOSURFACE_RGB10 1
#endif
```
The HDR images were displayed washed out when HDRForImagesEnabled is disabled on AS macOS. So I will keep this definition as is for now and I will test this patch on iOS to make sure the HDR images are converted to SDR correctly when HDRForImagesEnabled is disabled.

File: Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm (hunk 2 above)
Comment (shallawa) [reply Source/WebCore/platform/graphics/ca/cocoa/PlatformCALayerCocoa.mm]: `.wantsExtendedDynamicRangeContent = YES` is set for `ContentsFormat::RGBA10` and `ContentsFormat::RGBA16F`.

For WebGPU case we may need to add a new ContentsFormat enum value.

File: Source/WebCore/platform/graphics/cocoa/IOSurface.mm [hunk 7]
```diff
@@ -441,6 +462,10 @@ static IntSize computeMaximumSurfaceSize()
         break;
     case Format::RGBA:
         break;
+#if HAVE(HDR_SUPPORT)
+    case Format::RGBA16F:
+        break;
```
Comment (squelart) [inline Source/WebCore/platform/graphics/cocoa/IOSurface.mm]: Would it make sense here to copy what's in the "RGB10" case above? I.e.:
```
        bitsPerComponent = 16;
        bitmapInfo = static_cast<CGBitmapInfo>(kCGImageAlphaPremultipliedLast) | static_cast<CGBitmapInfo>(kCGBitmapByteOrder16Host) | static_cast<CGBitmapInfo>(kCGBitmapFloatComponents);
```

Comment (shallawa) [reply Source/WebCore/platform/graphics/cocoa/IOSurface.mm]: Yes. This makes sense.

**Pull request merged**